executor/src/witgen/generator.rs

```diff
@@ -105,27 +115,45 @@ where
             current_row_index: fixed_data.degree - 1,
             last_report: 0,
             last_report_time: Instant::now(),
-        };
-        // For identities like `pc' = (1 - first_step') * <...>`, we need to process the last
-        // row before processing the first row.
-        generator
-            .compute_next_row_or_initialize(generator.last_row(), ProcessingPhase::Initialization);
-        generator
+            initialized: false,
+        }
     }
 
     fn last_row(&self) -> DegreeType {
         self.fixed_data.degree - 1
     }
 
-    pub fn compute_next_row(&mut self, next_row: DegreeType) -> ColumnMap<T> {
-        self.compute_next_row_or_initialize(next_row, ProcessingPhase::Regular)
+    pub fn compute_next_row<Q>(
+        &mut self,
+        next_row: DegreeType,
+        mutable_state: &mut MutableState<'a, T, Q>,
+    ) -> ColumnMap<T>
+    where
+        Q: FnMut(&str) -> Option<T> + Send + Sync,
+    {
+        if !self.initialized {
+            // For identities like `pc' = (1 - first_step') * <...>`, we need to process the last
+            // row before processing the first row.
+            self.compute_next_row_or_initialize(
+                self.last_row(),
+                ProcessingPhase::Initialization,
+                mutable_state,
+            );
+
+            self.initialized = true
+        }
+        self.compute_next_row_or_initialize(next_row, ProcessingPhase::Regular, mutable_state)
     }
 
-    fn compute_next_row_or_initialize(
+    fn compute_next_row_or_initialize<Q>(
         &mut self,
         next_row: DegreeType,
         phase: ProcessingPhase,
-    ) -> ColumnMap<T> {
+        mutable_state: &mut MutableState<'a, T, Q>,
+    ) -> ColumnMap<T>
+    where
+        Q: FnMut(&str) -> Option<T> + Send + Sync,
```

Comment:
It's a bit annoying we have to add the query callback type parameter in so many places. An alternative would be to add it as a type parameter to `Generator`, but then `KnownMachines` would have to get the same parameter. But maybe that would make sense anyway, as there is no reason why a block machine couldn't make a query, for example?

Reply:
sounds good. We could also turn it into a proper trait.

Reply:
Hmm, I'd like to postpone this to another PR. Adding this type parameter to `Generator` propagates to types like `IdentityProcessor` (because it holds a list of Machines), `BlockMachine`, `Processor`, etc... See this commit: 675cc5d20a5aa9356e5f415a105b3b1839174a05